_posts/2019-12-10-bert-model.md

```diff
@@ -0,0 +1,410 @@
+```
+(v1.1) pkg> add TextAnalysis
+```
+
+Mange the versions as you'd want them. These versions *can* change in the future
+and you'll need to find one combination that works with all. Once, you've done
+that we can move towards coding part.
+
+You can find entire notebook here. (No link, will publish it after GCI).
+
+First, lets get all of the packages imported
+
+```jl
+using CuArrays
+
+using Transformers
+using Transformers.Basic
+using Transformers.Pretrain
+using Transformers.Datasets
+using Transformers.BidirectionalEncoder
+
+using Flux
+using Flux: onehotbatch, gradient
+import Flux.Optimise: update!
+using WordTokenizers
+```
+
+Also, some constants we'll be using:
+
+```jl
+const labels = ("0", "1")
+const opt = ADAM(1e-6)
+const Batch = 2
+const Epoch = 1
+```
+Of course, you can change these as per your requirement. I've set Batch to 2
+to save some memory and Epoch to 1 to save some time. Epoch being one just
+would mean that there will be a single iteration of weights update.
+
+Now, lets get the pretrained model. Here the downloading extracting the model
+will be taken care of by Transformers. All you need to do is following.
+
+```jl
+const _bert_model, wordpiece, tokenizer = pretrain"Bert-uncased_L-12_H-768_A-12"
+```
+
+There are many variants released for the pre-trained model which you can find
+on Transformers docs or online or just run `pretrains()` to find available ones.
+I'm going to go with the above one.
+
+You might need to type `Y` as stdin. Alternatively, you can set
+`DATADEPS_ALWAYS_ACCEPT=true` to avoid that.
+
+### Step 2: Preprocess data
+
+We need to first clean the data before we could use it. We can do this by using
+TextAnalysis.jl package.
+
+```jl
+using TextAnalysis
+
+function preprocess_data()
+    open("pre_train.txt") do file
+        lines = readlines(file)
+        for line in lines
+#             tag = line[1:10]
+            tag = line[1]
+            str_rev = line[3:end]
+            str_doc = StringDocument(str_rev)
+            prepare!(str_doc, strip_punctuation | strip_non_letters | strip_numbers)
+            remove_case!(str_doc)
+            stem!(str_doc)
+            text(str_doc)
+            open("new_pre_train.tsv", "a") do nfile
+                newtext = text(str_doc)
+                write(nfile, "$newtext\t$tag\n")
+            end
+        end
+    end
+end
+
+preprocess_data()
+```
+
+This above code might look a little ugly (I might need to learn few more
+functions but let's keep it as future work), but it gets the job done.
+You should be doing this for training data, since this will clean up most of it
+that we don't require.
+
+To prepare it the way we require, we would need to add few more functions. This
+will create an iterator like object for our data.
+
+```jl
+using DelimitedFiles
+
+function gen_train()
+    sets = readdlm("new_pre_train.tsv", '\t', String, quotes = false)
+    return [selectdim(sets, 2, i) for i = (1,2)]
```

Comment:
Is it possible to explain what this line does? what is the output you are looking for (beyond the fact that it's and iterator? Iterator over what? ie, what element does iterating over the thing return?